Add GitHub tests for karmadactl init --config command

tiansuo114 opened the pull request:
**What type of PR is this?**

<!--
Add one of the following kinds:

/kind api-change
/kind bug
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind feature
/kind flake

-->

**What this PR does / why we need it**:
This PR is an additional test, based on the suggestion provided in [url](https://github.com/karmada-io/karmada/pull/5357#discussion_r1807568543).
**Which issue(s) this PR fixes**:
Part of #5387
It adds test cases for deploying Karmada services using the `karmadactl init --config` command with a configuration file, similar to the existing tests for `karmadactl init`.

**Special notes for your reviewer**:
@liangyuanpeng 
@RainbowMango 

~~This PR currently contains two commits because the test files depend on the new feature for deploying with a configuration file, introduced in [#5357](https://github.com/karmada-io/karmada/pull/5357). Once that PR is merged, this PR will be rebased, retaining only the commit for adding test cases. However, to save time, it is being submitted for review in advance.~~

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required.
-->
```release-note

```

tiansuo114 commented:
May need help😭😭, in my own project's github action, these tests passed correctly, but when I commit, these tests become failed, what should I do to fix these problems
@liangyuanpeng

liangyuanpeng commented:
/assign 
/retest

liangyuanpeng commented:
/ok-to-test

liangyuanpeng commented on pkg/karmadactl/cmdinit/kubernetes/deploy.go:
```diff
@@ -729,3 +730,47 @@ func generateServerURL(serverIP string, nodePort int32) (string, error) {
+func NewDefaultCommandInitOption() *CommandInitOption {
+	return &CommandInitOption{
+		ImagePullPolicy:                    string(corev1.PullIfNotPresent),
+		KubeImageTag:                       "v1.29.6",
+		Namespace:                          "karmada-system",
```
does we have not constants for it?

tiansuo114 replied:
I was not observed in the search system, which has the corresponding constants, in NewDefaultCommandInitOption approach is to write, I mainly refer to is the PKG/karmadactl/cmdinit/cmdinit. Go in for the default value of cmdflag, there are Parts of const variables are mostly copied using const constants

Will need these for example KubeImageTag string constant changes in the form of variables on the PKG/karmadactl/cmdinit/kubernets/deploy. Go in?

zhzhuang-zju replied:
> Will need these for example KubeImageTag string constant changes in the form of variables on the PKG/karmadactl/cmdinit/kubernets/deploy. Go in?

How about creating constants like `KubeImageTag` and `KarmadaNamespace` to reduce shotgun surgery modifications?

tiansuo114 replied:
> > Will need these for example KubeImageTag string constant changes in the form of variables on the PKG/karmadactl/cmdinit/kubernets/deploy. Go in?将需要将这些例如 KubeImageTag 字符串常量以变量的形式放在 PKG/karmadactl/cmdinit/kubernets/deploy 上。进去？
> 
> How about creating constants like `KubeImageTag` and `KarmadaNamespace` to reduce shotgun surgery modifications?创建像 `KubeImageTag` 和 `KarmadaNamespace` 这样的常量来减少霰弹枪手术修改怎么样？

Ok, I'll add some const variables to the config package and modify the code

liangyuanpeng commented on pkg/karmadactl/config/init/list_option.go:
```diff
@@ -0,0 +1,55 @@
+	return nil
+}
+
+// Run start list Karmada addons
```
Seems like this comment is nocorrect.

liangyuanpeng commented on pkg/karmadactl/config/init/list_option.go:
```diff
@@ -0,0 +1,55 @@
+	"github.com/karmada-io/karmada/pkg/karmadactl/cmdinit/kubernetes"
+)
+
+// CommandConfigImageOption options for addons list.
```
```suggestion
// CommandConfigImageOption options for components.
```

liangyuanpeng reviewed (commented):
Thanks you working for it, could you please add some e2e test here?  Maybe take a look at how kubeadm does it.

tiansuo114 commented:
> Thanks you working for it, could you please add some e2e test here? Maybe take a look at how kubeadm does it.

I have fixed some bugs and added e2e tests, please let me know if you need to make further changes

tiansuo114 commented:
> Thanks you working for it, could you please add some e2e test here? Maybe take a look at how kubeadm does it.

I have a question regarding the construction of the karmadactl application in the E2E test environment. While observing the workflow in the setup e2e test environment section, I did not find any actions to compile it using the Makefile. Without using the Makefile for compilation, the -X github.com/karmada-io/karmada/pkg/version.gitVersion=v1.11.0-alpha.0-254-gf3bcb21ff-dirty linker flag is not applied, which might be the reason why the output of karmadactl config image list in the E2E test is different from my expected output. How should I solve this issue?

tiansuo114 commented:
> Thanks you working for it, could you please add some e2e test here? Maybe take a look at how kubeadm does it.感谢您的工作，您可以在此处添加一些 e2e 测试吗？也许看看 kubeadm 是如何做到的。

"I have fixed the issues you pointed out and provided new e2e test cases. Please let me know if there are any further modifications needed. :)

liangyuanpeng commented:
> This PR is related to the documentation PR: https://github.com/karmada-io/karmada/pull/5277

@tiansuo114 

Add to PR issue body would be great, thank you! 

![image](https://github.com/user-attachments/assets/3b46e55c-da53-4f5a-a08d-05308dabc88f)

tiansuo114 commented:
> > This PR is related to the documentation PR: #5277此 PR 与文档 PR 相关： #5277
> 
> @tiansuo114
> 
> Add to PR issue body would be great, thank you!添加到 PR 问题正文会很棒，谢谢！
> 
> ![image](https://private-user-images.githubusercontent.com/28711504/357079664-3b46e55c-da53-4f5a-a08d-05308dabc88f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1MzM4NjQsIm5iZiI6MTcyMzUzMzU2NCwicGF0aCI6Ii8yODcxMTUwNC8zNTcwNzk2NjQtM2I0NmU1NWMtZGE1My00ZjVhLWEwOGQtMDUzMDhkYWJjODhmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEzVDA3MTkyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY4NWNiY2U3ZTEzYjVhMzA2NGQ4OWViODg3Y2EyMjJmYjNjZmIzMGViYzE3ZDk0NGZjNTA3NTY4ODcyM2M0YzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1xHhbFvuoDtKZpvqJk9jxz4mvBOkyNntCrmxGT_6aY0)

Ok, I modified the pr and the other two questions about pr and added links to it

XiShanYongYe-Chang commented:
Ask @zhzhuang-zju have a look.
/assign @zhzhuang-zju

zhzhuang-zju commented on pkg/karmadactl/karmadactl.go:
```diff
@@ -96,6 +97,7 @@ func NewKarmadaCtlCommand(cmdUse, parentCommand string) *cobra.Command {
 			Commands: []*cobra.Command{
 				explain.NewCmdExplain(f, parentCommand, ioStreams),
 				get.NewCmdGet(f, parentCommand, ioStreams),
+				config.NewCmdConfig(parentCommand),
```
The `config` command serves the `init` command, so it would be better to place it in the "Cluster Registration Commands:" group.

zhzhuang-zju commented on pkg/karmadactl/cmdinit/kubernetes/deploy.go:
```diff
@@ -729,3 +730,47 @@ func generateServerURL(serverIP string, nodePort int32) (string, error) {
+		KarmadaSchedulerReplicas:           1,
+		KarmadaControllerManagerImage:      DefaultKarmadaControllerManagerImage,
+		KarmadaControllerManagerReplicas:   1,
+		KarmadaWebhookImage:                DefaultKarmadaWebhookImage,
+		KarmadaWebhookReplicas:             1,
+		KarmadaAggregatedAPIServerImage:    DefaultKarmadaAggregatedAPIServerImage,
+		KarmadaAggregatedAPIServerReplicas: 1,
+		WaitComponentReadyTimeout:          cmdinitoptions.WaitComponentReadyTimeout,
+		CertValidity:                       cert.Duration365d,
+		HostClusterDomain:                  globaloptions.DefaultHostClusterDomain,
+	}
```
func `NewDefaultCommandInitOption` is intended for use with the `config` command, but the `config` command actually only uses a few of the options. Is it necessary to set default values for all options?

tiansuo114 replied:
I think it could be used as a general initialization operation to facilitate the addition of other commands related to karmadactl config in the future. Adding all the default information here might be helpful for future development work, but removing this information and adding it only when needed is also an option. Which approach do you think is better?

zhzhuang-zju replied:
> Adding all the default information here might be helpful for future development work, but removing this information and adding it only when needed is also an option. Which approach do you think is better?

I perfer the latter. If subsequent development requires a complete default initialization operation, we can extract the operations from `flags.XXXVarP` to obtain a generic initialization operation that includes default values.

tiansuo114 replied:
> > Adding all the default information here might be helpful for future development work, but removing this information and adding it only when needed is also an option. Which approach do you think is better?
> 
> I perfer the latter. If subsequent development requires a complete default initialization operation, we can extract the operations from `flags.XXXVarP` to obtain a generic initialization operation that includes default values.

I have fixed the defects mentioned above, how does the code look now

liangyuanpeng commented:
/hold 
need squash the commits when ready to merge.

tiansuo114 commented:
> /hold /抓住 need squash the commits when ready to merge.准备合并时需要压缩提交。

get,Ok, I will squash commit after the problem in comment is solved

tiansuo114 commented:
> > /hold /抓住 need squash the commits when ready to merge.准备合并时需要压缩提交。
> 
> get,Ok, I will squash commit after the problem in comment is solvedget，好的，评论中的问题解决后我会squash commit

tiansuo114 commented on pkg/karmadactl/config/image.go:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	imageParentCommand := fmt.Sprintf("%s %s", parentCommand, "image")
+	cmd.AddCommand(NewCmdConfigImagesList(imageParentCommand))
```
Now that the comments have been fixed, how does the code look now

zhzhuang-zju commented on pkg/karmadactl/config/list.go:
```diff
@@ -0,0 +1,59 @@
+// NewCmdConfigImagesList creates a new list command
+func NewCmdConfigImagesList(parentCommand string) *cobra.Command {
+	opts := configInit.CommandConfigImageOption{
+		InitOption: kubernetes.NewDefaultCommandInitOption(),
```
Could we consider removing the function `NewDefaultCommandInitOption`, as default values can all be set through `cmd.Flags().XXXVarP`?
And if testing requires it, we could also move `NewDefaultCommandInitOption` to the corresponding test file.

liangyuanpeng replied:
@tiansuo114  could you explain why we need it?

tiansuo114 replied:
> Could we consider removing the function `NewDefaultCommandInitOption`, as default values can all be set through `cmd.Flags().XXXVarP`?我们是否可以考虑删除函数 `NewDefaultCommandInitOption` ，因为默认值都可以通过 `cmd.Flags().XXXVarP` 设置？ And if testing requires it, we could also move `NewDefaultCommandInitOption` to the corresponding test file. 如果测试需要，我们也可以将 `NewDefaultCommandInitOption` 移动到相应的测试文件中。

I am attempting to discard the NewDefaultCommandInitOption method and use the following method to create CommandInitOption in the file.
```
func NewCmdConfigImagesList(parentCommand string) *cobra.Command {
	opts := configInit.CommandConfigImageOption{
		InitOption: &kubernetes.CommandInitOption{},
	}
	cmd := &cobra.Command{
```
However, when I actually execute it, the command produces the following error.
```
registry.k8s.io/kube-apiserver:
registry.k8s.io/kube-controller-manager:
registry.k8s.io/etcd:3.5.13-0



```
When the NewDefaultCommandInitOption initialization is not performed, the output string related to the image will be empty.
I haven't found a good way to solve this issue, so I have continued to keep the NewDefaultCommandInitOption method in use.

zhzhuang-zju replied:
So, the `karmadactl config image list` command currently does not support configuring image tags, correct?

tiansuo114 replied:
In what is now the code is to support the image of the tag, the image of the tags related to fill github.com/karmada-io/karmada/pkg/version.gitVersion fields at compile time

tiansuo114 replied:
@zhzhuang-zju 
I have added the output that the command would normally provide at the beginning of the PR. Regarding the question in the comment about the NewDefaultCommandInitOption method, what do you think would be the most appropriate way to handle this? Should we keep the currently working method, or do you have a better solution to suggest?

liangyuanpeng replied:
> So, the command currently does not support configuring image tags, correct?karmadactl config image list

how about let's reduce review pressure to quickly move forward with PR consolidation, leave it to the next PR.

I think you're talking about the flag of `kube-image-tag` (we have not karmada-imag-tag yet)

zhzhuang-zju replied:
> I think you're talking about the flag of kube-image-tag (we have not karmada-imag-tag yet)

Yes, I think it would be much better to introduce the flag `kube-image-tag`. 
@tiansuo114  Could you help confirm whether the function `NewDefaultCommandInitOption` will still be needed after introducing the flag `kube-image-tag`? Thanks.

tiansuo114 replied:
@zhzhuang-zju 
In my understanding, the output of the config image list command consists of three parts:
repository link + image name + image tag
In this command, the first two parts rely on the initialization of kubernetes.CommandInitOption. Therefore, even after introducing kube-image-tag, the NewDefaultCommandInitOption method is still necessary.

zhzhuang-zju replied:
Thank you for your answer. If there are no current plans to expose more configurable flags, the current implementation is fine with me.

tiansuo114 replied:
get,I will fix the remaining issues as soon as possible and prepare for the next step

zhzhuang-zju commented on pkg/karmadactl/config/list.go:
```diff
@@ -0,0 +1,59 @@
+
+var (
+	listExample = templates.Examples(`
+	# List Karmada all required images from the specified private image registry
```
@tiansuo114 The original sentence is not very smooth; it would be better if it could be modified~

liangyuanpeng commented on test/e2e/karmadactl_test.go:
```diff
@@ -1222,6 +1222,87 @@ var _ = ginkgo.Describe("Karmadactl options testing", func() {
+
+		ginkgo.It("list images with kube-image-country flag", func() {
+			cmdArgs = []string{"config", "image", "list", "--kube-image-country=cn"}
+			expected = `registry.cn-hangzhou.aliyuncs.com/google_containers/kube-apiserver:v1.29.6
```
Once you rebase PR, this will need to be 1.30.4 and not 1.29.6

liangyuanpeng commented on pkg/karmadactl/cmdinit/options/global.go:
```diff
@@ -40,3 +40,20 @@ const (
+
+const (
+	// DefaultKubeImageTag is the default tag for the Kubernetes image used by the Karmada control plane.
+	DefaultKubeImageTag = "v1.29.6"
```
```suggestion
	DefaultKubeImageTag = "v1.30.4"
```

tiansuo114 replied:
have fixed and rebased

liangyuanpeng commented:
@zhzhuang-zju  anyelse comment here?

tiansuo114 commented:
> @zhzhuang-zju anyelse comment here?

Can this pr be merged now or does it need to be further modified😭😭

zhzhuang-zju commented:
/lgtm
assign @XiShanYongYe-Chang

tiansuo114 commented:
The command in this feature PR has been modified to match the [current documentation](https://github.com/karmada-io/karmada/pull/5623) format, but its functionality remains unchanged and is still the same as it was initially.
@liangyuanpeng @RainbowMango

tiansuo114 commented:
Based on community meetings, it seems that this feature is not necessary and will increase maintenance costs, so this pr is temporarily closed

liangyuanpeng commented:
https://github.com/karmada-io/karmada/pull/5744 is migration to here (for Summer OSPP 2024)

tiansuo114 commented:
Based on the OSPP project's requirements, this PR has been reopened. However, the relevant content in this PR has been changed to add automated testing for the new `karmadactl init --config` command as indicated in the [link](https://github.com/karmada-io/karmada/pull/5744).

@liangyuanpeng @RainbowMango @zhzhuang-zju

tiansuo114 commented:
/hold cancel

tiansuo114 commented:
> #5744 is migration to here (for Summer OSPP 2024)

The code in this PR has been re-pushed and tested, and after a code clean-up, I’ve removed the unnecessary metadata part in KarmadaInitConfig. We can now continue with the previous review.🎉

If need to track this [issue](https://github.com/karmada-io/karmada/pull/5357#discussion_r1818713706), please refer to the relevant changes in this PR. 
@RainbowMango @liangyuanpeng

liangyuanpeng reviewed (commented):
thank you!
/lgtm 
/cc @zhzhuang-zju  if you want to review for karmadactl

RainbowMango commented on .github/workflows/cli.yaml:
```diff
@@ -62,3 +62,45 @@ jobs:
+        uses: actions/upload-artifact@v4
+        with:
+          name: karmadactl_config_test_logs_${{ matrix.k8s }}
+          path: ${{ github.workspace }}/karmadactl-test-logs/${{ matrix.k8s }}/config/
```
end of line.

tiansuo114 replied:
Do you mean it’s best to add a blank line at the end of the line? If so, the code has already been updated. If I’ve misunderstood, please feel free to correct me.

liangyuanpeng replied:
> Do you mean it’s best to add a blank line at the end of the line? 

correct

RainbowMango reviewed (commented):
Looks good to me, just a nit, otherwise LGTM.

RainbowMango reviewed (commented):
/lgtm
/assign @liangyuanpeng

RainbowMango commented:
It seems this is the last PR for the summer OSPP program. #5387.
Please let me know if I miss anything.

tiansuo114 commented:
> It seems this is the last PR for the summer OSPP program. #5387. Please let me know if I miss anything.


Yes, this PR is the last one for the OSPP tasks. Up until now, with the merge of this PR into the main branch, all planned PRs have been incorporated into Karmada.
PR Link 1: https://github.com/karmada-io/karmada/pull/5277
PR Link 2: https://github.com/karmada-io/karmada/pull/5306
PR Link 3: https://github.com/karmada-io/karmada/pull/5357
PR Link 4: https://github.com/karmada-io/website/pull/690
PR Link 5: https://github.com/karmada-io/karmada/pull/5321

liangyuanpeng commented:
> It seems this is the last PR for the summer OSPP program. https://github.com/karmada-io/karmada/issues/5387.

Yes, it is.
/lgtm
/approve

RainbowMango reviewed (approved):
/approve
